File: src/main/java/toby/command/commands/MoveCommand.java
package toby.command.commands;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.*;
import toby.command.CommandContext;
import toby.command.ICommand;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class MoveCommand implements ICommand {

    public static Long badOpinionChannel = 756262044491055165L;

    @Override
    public void handle(CommandContext ctx, String prefix) {
        final TextChannel channel = ctx.getChannel();
        final Message message = ctx.getMessage();
        final Member member = ctx.getMember();
        final List<String> args = ctx.getArgs();
        VoiceChannel voiceChannel;
        Guild guild = ctx.getGuild();

        if (message.getMentionedMembers().isEmpty()) {
            channel.sendMessage("You must mention 1 or more Users to move").queue();
            return;
        }

        Optional<VoiceChannel> voiceChannelOptional;
        try {
            voiceChannelOptional = args.stream()
                    .filter(s -> !s.matches(Message.MentionType.USER.getPattern().pattern()))
                    .map(guild::getVoiceChannelById)
                    .collect(Collectors.toList()).stream().findFirst();
        } catch (Error e) {
            return;
        }
        voiceChannel = voiceChannelOptional.orElseGet(() -> guild.getVoiceChannelById(badOpinionChannel));
        message.getMentionedMembers().forEach(target -> {

            if (doChannelValidation(ctx, channel, member, target)) return;

            guild.moveVoiceMember(target, voiceChannel)
                    .queue(
                            (__) -> channel.sendMessageFormat("Moved %s to '%s'", target.getEffectiveName(), voiceChannel.getName()).queue(),
                            (error) -> channel.sendMessageFormat("Could not move '%s'", error.getMessage()).queue()
                    );
        });
    }

    private boolean doChannelValidation(CommandContext ctx, TextChannel channel, Member member, Member target) {
        if(!target.getVoiceState().inVoiceChannel()){
            channel.sendMessage(String.format("Mentioned user '%s' is not connected to a voice channel currently, so cannot be moved.", target.getEffectiveName())).queue();
            return true;
        }
        if (!member.canInteract(target) || !member.hasPermission(Permission.VOICE_MOVE_OTHERS)) {
            channel.sendMessage(String.format("You can't move '%s'", target.getEffectiveName())).queue();
            return true;
        }

        final Member botMember = ctx.getSelfMember();

        if (!botMember.hasPermission(Permission.VOICE_MOVE_OTHERS)) {
            channel.sendMessage(String.format("I'm not allowed to move %s", target.getEffectiveName())).queue();
            return true;
        }
        return false;
    }

    @Override
    public String getName() {
        return "move";
    }

    @Override
    public String getHelp(String prefix) {
        return "move a member into a voice channel.\n" +
                String.format("Usage: `%smove <@user> channelId (right click voice channel and copy id)`\n", prefix) +
                "e.g. !move @username 756262044491055165";
    }
}

File: discord-bot/src/main/kotlin/bot/toby/command/commands/moderation/MoveCommand.kt
package bot.toby.command.commands.moderation

import core.command.Command.Companion.replyAndDelete
import core.command.CommandContext
import database.dto.ConfigDto
import database.dto.UserDto
import database.service.ConfigService
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class MoveCommand @Autowired constructor(private val configService: ConfigService) : ModerationCommand {

    companion object {
        private const val USERS = "users"
        private const val CHANNEL = "channel"
    }

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int) {
        val event = ctx.event
        event.deferReply().queue()

        val guild = event.guild ?: return
        val member = ctx.member ?: return
        val botMember = guild.selfMember

        val memberList = event.getOption(USERS)?.mentions?.members.orEmpty()
        if (memberList.isEmpty()) {
            event.hook.replyAndDelete("You must mention 1 or more Users to move", deleteDelay)
            return
        }

        val voiceChannel = getVoiceChannel(event, guild) ?: run {
            event.hook.replyAndDelete("Could not find a channel on the server that matched the name", deleteDelay)
            return
        }

        memberList.forEach { target ->
            if (!validateChannel(event, botMember, member, target, deleteDelay)) {
                guild.moveVoiceMember(target, voiceChannel).queue(
                    {
                        event.hook.replyAndDelete(
                            "Moved ${target.effectiveName} to '${voiceChannel.name}'",
                            deleteDelay,
                        )
                    },
                    { error ->
                        event.hook.replyAndDelete(
                            "Could not move '${target.effectiveName}': ${error.message}",
                            deleteDelay,
                        )
                    }
                )
            }
        }
    }

    private fun getVoiceChannel(event: SlashCommandInteractionEvent, guild: Guild) =
        event.getOption(CHANNEL)?.asChannel?.asVoiceChannel()
            ?: configService.getConfigByName(ConfigDto.Configurations.MOVE.configValue, guild.id)
                ?.let { config -> guild.getVoiceChannelsByName(config.value ?: "", true).firstOrNull() }

    private fun validateChannel(
        event: SlashCommandInteractionEvent,
        botMember: Member,
        member: Member,
        target: Member,
        deleteDelay: Int
    ): Boolean {
        return when {
            target.voiceState?.inAudioChannel() == false -> {
                event.hook.replyAndDelete(
                    "Mentioned user '${target.effectiveName}' is not connected to a voice channel currently, so cannot be moved.",
                    deleteDelay,
                )
                true
            }
            !member.canInteract(target) || !member.hasPermission(Permission.VOICE_MOVE_OTHERS) -> {
                event.hook.replyAndDelete("You can't move '${target.effectiveName}'", deleteDelay)
                true
            }
            !botMember.hasPermission(Permission.VOICE_MOVE_OTHERS) -> {
                event.hook.replyAndDelete("I'm not allowed to move ${target.effectiveName}", deleteDelay)
                true
            }
            else -> false
        }
    }

    override val name: String
        get() = "move"

    override val description: String
        get() = "Move mentioned members into a voice channel (voice channel can be defaulted by config command)"

    override val optionData: List<OptionData>
        get() = listOf(
            OptionData(OptionType.STRING, USERS, "User(s) to move", true),
            OptionData(OptionType.STRING, CHANNEL, "Channel to move to")
        )
}
